src/tool/hpcrun/gpu/nvidia/cupti-api.c

```diff
@@ -570,29 +601,41 @@ cuda_path
 static void
 cupti_set_default_path(char *buffer)
 {
-  strcpy(buffer, CUPTI_INSTALL_PREFIX CUPTI_LIBRARY_LOCATION);
+  strcpy(buffer, CUPTI_INSTALL_PREFIX "/" CUPTI_LIBRARY_LOCATION);
 }
 
-int
+
+static int
 library_path_resolves(const char *buffer)
 {
   struct stat sb;
   return stat(buffer, &sb) == 0;
 }
 
 
-static const char *
+const char *
 cupti_path
 (
-  void
+ void
 )
 {
   const char *path = "libcupti.so";
-  int resolved = 0;
 
   static char buffer[PATH_MAX];
   buffer[0] = 0;
 
+#ifdef NEW_CUPTI
+  // XXX(Keren): Don't use the default cupti library under CUDA_HOME because
+  // since NVIDIA delivers us independent cupti libraries that contain bug
+  // fixes.
```

Comment:
Wouldn't this cause issues between mismatching CUDA and CUPTI? If it doesn't, why do we mess about in the old-CUPTI case with finding the CUDA install path at all?

Reply:
1. It does not cause issuing between CUDA and CUPTI. CUPTI > 11.0 can be used with any CUDA > 11.0. 
2. It is put here for now just to test nvidia's CUPTI patches. In principal, the correct CUPTI library should come with the CUDA library. However, we are using standalone CUPTI libraries just for testing. In deployment, we don't link CUPTI library outside of the CUDA directory.

Reply:
> CUPTI > 11.0 can be used with any CUDA > 11.0.

Is that true in practice? I vaguely remember a case where we couldn't get any data out because the two had a version mismatch within the 11.x series.

If it is compatible, I would think it easier to just always use the one we link with and error if the loaded CUDA isn't a compatible version (which AFAIK shouldn't happen for Spack-built Toolkit since it messes with the RPATHs... which is also a complicating factor here). That way we don't have to worry about the headers changing in backward-incompatible ways between the major version series (because honestly I don't trust NVIDIA not to botch it somehow).

Either way, I would really prefer a different activation conditional here, having it bound to `--enable-new-cupti` doesn't make sense to me yet. Maybe an environment var/control knob/`-dd` to skip the search for the CUPTI bundled with CUDA?

Reply:
> Is that true in practice? I vaguely remember a case where we couldn't get any data out because the two had a version mismatch within the 11.x series.

Well, this is how CUPTI is designed to be. After CUDA 11.0, these CUDA libraries should have good forward compatibility and backward compatibility among major versions (11.x). I admit that there were still minor issues though.

> Either way, I would really prefer a different activation conditional here, having it bound to --enable-new-cupti doesn't make sense to me yet. Maybe an environment var/control knob/-dd to skip the search for the CUPTI bundled with CUDA?

Let me rephrase my sentences here. I don't want this special cupti binding at all, and it makes no sense to link a CUPTI library with a CUDA library with different versions. It is put here just because CUPTI is buggy at this moment so that we cannot use the default CUPTI library shipped with CUDA. 

And the whole point of using NEW_CUPTI is to have a temporary version that tests the new pc sampling api. Once we could confirm that CUPTI does not have significant problems, I'll remove the NEW_CUPTI macros and merge similar code.